### PR TITLE
chore(Chat): move to functional component

### PR DESCRIPTION
**Pull request by layershifter**
# BREAKING CHANGES

As in other components that are moved functional we limited sets of props that are passed to styles functions.

| `Chat`              |
| ------------------- |
| no props are passed |

| `ChatItem`        |
| ----------------- |
| `attached`        |
| `contentPosition` |

| `ChatMessage`                                        |
| ---------------------------------------------------- |
| `attached`                                           |
| `badgePosition`                                      |
| `mine`                                               |
| `focused`                                            |
| `hasBadge` (replacement for `badge`)                 |
| `hasReactionGroup` (replacement for `reactionGroup`) |

### CODE REVIEW

File: packages/react/src/components/Chat/ChatMessage.tsx [hunk 1]
```diff
@@ -113,116 +121,118 @@ export interface ChatMessageProps
...
-
-  updateActionsMenuPosition = () => {}
-
-  state = {
-    focused: false,
-    messageNode: null,
-  }
-
-  menuRef = React.createRef<HTMLElement>()
+export type ChatMessageStylesProps = Pick<
```
Comment (mnajdova) [inline packages/react/src/components/Chat/ChatMessage.tsx:124]: It's good that we have these typings now on the component level, but let's make sure that we move the others too :)

File: packages/react/src/components/Chat/ChatMessage.tsx [hunk 2]
```diff
@@ -113,116 +121,118 @@ export interface ChatMessageProps
...
+      }),
+      rtl: context.rtl,
     },
-  }
+  )
 
-  handleFocus = (e: React.SyntheticEvent) => {
-    this.updateActionsMenuPosition()
+  const handleFocus = (e: React.SyntheticEvent) => {
+    if (updateActionsMenuPosition.current) updateActionsMenuPosition.current()
```
Comment (mnajdova) [inline packages/react/src/components/Chat/ChatMessage.tsx]: `_.invoke(updateActionsMenuPosition, 'current')`?

File: packages/react/src/components/Chat/ChatMessage.tsx [hunk 3]
```diff
@@ -113,116 +121,118 @@ export interface ChatMessageProps
...
-  handleMouseEnter = (e: React.SyntheticEvent) => {
-    this.updateActionsMenuPosition()
-    _.invoke(this.props, 'onMouseEnter', e, this.props)
+    setFocused(shouldPreserveFocusState)
+    _.invoke(props, 'onBlur', e, props)
   }
 
-  handleMessageRef = (node: HTMLElement) => this.setState({ messageNode: node })
+  const handleMouseEnter = (e: React.SyntheticEvent) => {
+    if (updateActionsMenuPosition.current) updateActionsMenuPosition.current()
```
Comment (mnajdova) [inline packages/react/src/components/Chat/ChatMessage.tsx]: Same as above

File: packages/react/src/components/Chat/ChatItem.tsx [hunk 4]
```diff
@@ -45,108 +55,149 @@ export interface ChatItemProps extends UIComponentProps, ChildrenComponentProps
...
     return messageElement
   }
 
-  cloneElementWithCustomProps = (element, props, prop?) => {
-    if (!prop) {
-      return React.cloneElement(element, props)
-    }
-    return React.cloneElement(element, {
-      [prop]: React.cloneElement(getElementProp(element, prop), props),
+  const renderChatItem = () => {
```
Comment (mnajdova) [inline packages/react/src/components/Chat/ChatItem.tsx]:
```suggestion
  const renderContent = () => {
```

File: packages/react/src/components/Chat/ChatItem.tsx [hunk 5]
```diff
@@ -45,108 +55,149 @@ export interface ChatItemProps extends UIComponentProps, ChildrenComponentProps
...
+
+  const element = (
+    <ElementType
+      {...getA11Props('root', {
+        className: classes.root,
+        ...rtlTextContainer.getAttributes({ forElements: [children] }),
+        ...unhandledProps,
+      })}
+    >
+      {childrenExist(children) ? children : renderChatItem()}
```
Comment (mnajdova) [inline packages/react/src/components/Chat/ChatItem.tsx]:
```suggestion
      {childrenExist(children) ? children : renderContent()}
```

File: packages/react/src/components/Chat/Chat.tsx [hunk 6]
```diff
@@ -1,80 +1,126 @@
...
-  accessibility?: Accessibility
+  accessibility?: Accessibility<ChatBehaviorProps>
 
   /** Shorthand array of the items inside the chat. */
   items?: ShorthandCollection<ChatItemProps>
 }
 
-class Chat extends UIComponent<WithAsProp<ChatProps>, any> {
-  static displayName = 'Chat'
+export type ChatStylesProps = {}
```
Comment (mnajdova) [inline packages/react/src/components/Chat/Chat.tsx:46]: Yeeey!

File: packages/react/src/components/Chat/Chat.tsx [hunk 7]
```diff
@@ -1,80 +1,126 @@
...
+        className: classes.root,
+        ...rtlTextContainer.getAttributes({ forElements: [children] }),
+        ...unhandledProps,
+      })}
+    >
+      {childrenExist(children)
+        ? children
+        : _.map(items, item =>
+            ChatItem.create(item, {
+              defaultProps: () => ({ className: Chat.slotClassNames.item }),
```
Comment (mnajdova) [inline packages/react/src/components/Chat/Chat.tsx:90]: Shouldn't we by convention send here `resolvedStyles.item`? I hope not, but still, thought I'd ask

Comment (layershifter) [reply packages/react/src/components/Chat/Chat.tsx:90]: Discussed offline: let's avoid changes there and keep previous behavior.

File: packages/react/src/components/Chat/ChatMessage.tsx (hunk 3 above)
Comment (layershifter) [reply packages/react/src/components/Chat/ChatMessage.tsx]: Done 👍

File: packages/react/src/components/Chat/ChatMessage.tsx [hunk 8]
```diff
@@ -266,85 +275,72 @@ class ChatMessage extends UIComponent<WithAsProp<ChatMessageProps>, ChatMessageS
...
-
-    return (
-      <Ref innerRef={this.handleMessageRef}>
+  const childrenPropExists = childrenExist(children)
+  const rootClasses = childrenPropExists ? cx(classes.root, classes.content) : classes.root
+
+  const ElementType = getElementType(props)
+  const unhandledProps = getUnhandledProps(ChatMessage.handledProps, props)
+
+  const badgeElement = Label.create(badge, {
```
Comment (mnajdova) [inline packages/react/src/components/Chat/ChatMessage.tsx:292]: Another example that we were missing `getA11Props` :(

**Review by mnajdova — APPROVED** [review]

**Pull request merged**